Content update throughout payment channel sub-section (#1004)
filecoin-project/specs: content/systems/filecoin_token/payment_channels/_index.md

title: Payment Channels
weight: 2
bookCollapseSection: true
dashboardAudit: 1
dashboardState: Stable
dashboardInterface: wip
---

# Payment Channels
---

Payment Channels are used in the Filecoin [Retrieval Market](retrieval_market) to enable efficient off-chain payments and accounting between parties for what is expected to be a series of microtransactions, as these occur during data retrieval.

> In particular, given that there is no proving method provided for the act of sending data from a provider (miner) to a client, there is no trust anchor between the two. Therefore, in order to avoid mis-behaviour, Filecoin is making use of payment channels in order to realise a step-wise "data transfer <-> payment" relationship between the provider and the recipient of data. Clients issue requests for data that miners are responding to. The miner is entitled to ask for interim payments, the volume-oriented interval for which is agreed in the Deal phase. In order to facilitate this process, the Filecoin client is creating a payment channel once the provider has agreed on the proposed deal. The client should also lock in the payment channel monetary value equal to the one needed for retrieval of the entire block of data requested. Every time a provider is completing transfer of the pre-specified amount of data, they can request a payment. The client is responding to this payment with a voucher which the provider can redeem (immediately or later).

Reviewer: So, I am a bit ambivalent about including this paragraph here.

It's important to note that while Payment Channels are primarily used by Retrieval in the software we've written they are also an independent feature, and they are intended to be a platform for people to write applications that involve payment transfers on top of Filecoin. 

I think the purpose of payment channels is better summarized below:

"The goal of the payment channel actor is to enable a series of off-chain microtransactions to be reconciled on-chain at a later time with fewer messages."

Anyway, this is helpful to understanding how payment and retrieval work together, but I would maybe make a seperate section called "Payment channels and retrieval" to emphasize that payment channels are a feature, and retrieval market is one major use case.

> Redeeming a voucher is not transferring funds from the payment channel to the provider's account, that is, it is not an on-chain operation. Instead, redeeming a voucher denotes the fact that data-volume worth of `voucher_val` has been transferred to the client and the corresponding value is locked on the provider's side of payment channel. It is not until the voucher is submitted on-chain through `Collect` that the funds are dispatched to the provider's account.

Reviewer: "Redeeming a voucher is not transferring funds from the payment channel to the provider's account"

True -- no transfer happens until Collect is called.

"that is, it is not an on-chain operation."

False, or at least confusing in my mind.

There are three on-chain operations:
UpdateChannelState = how you submit vouchers
Settle - to begin closing a channel
Collect- When funds are actually transfered from client to provider

Redeem in my mind means submitting vouchers to the chain through UpdateChannelState

I think this entire paragraph is ultimately confusing.

Reviewer: the process is:
Client creates a voucher -> sends to Provider
Provider -> saves the voucher locally
Either immediately or later, the Provider "redeems" the voucher by submitting it to the chain, calling UpdateChannelState (in the current software, it's always later -- our retrieval code does not call UpdateChannelState -- it has to be done manually from the CLI)
Client or Provider -> Calls settle
12 hour period begins
If the provider has not already done so, they should now submit the vouchers to the chain (Lotus now does this automatically)
12 hour period ends
Client or Provider -> calls Collect
Funds are transferred to Provider, any unclaimed balance goes back to Client

In addition, payment channels in Filecoin can be split into `lane`s created as part of updating the channel state with a payment `voucher`. Each lane has an associated `nonce` and amount of tokens it can be `redeemed` for. These lanes allow for a lot of accounting between parties to be done off-chain and reconciled via single updates to the payment channel.

One example of such accounting is *merging of lanes*. When a pair of client-provider nodes have a payment channel established between them with many lanes, the provider will have to pay gas cost for each one of the lanes in order to `Collect` funds. Alternatively, the provider can send a "merge" request to the client to ask it to merge some of the lanes and consolidate the funds. As an incentive for the client to accept the merge lane request, the provider can ask for lower fees to balance out the gas cost. For example, if the provider has collected vouchers worth of 10 monetary units from two lanes, say 5 from each, and the gas cost is 2, then it can ask for 9 from the client if the latter accepts to merge the two lanes.

> Over the course of a transaction cycle, each participant in the payment channel can send `voucher`s to other participants. The payment channel's client (`From`) account holder will send a signed voucher with a given nonce to the provider (`To`) account holder.

Reviewer: I would lean away from using the terms `client` and `provider` here as they are retrieval specific. I would use `creator` (instead of client) and `recipient` (instead of provider)

The payment channel state structure looks like this:

```go
// A given payment channel actor is established by From (normally the client/receipent of data)
// to enable off-chain microtransactions to To (normally provider/sender of data) to be reconciled
// and tallied on chain.
type State struct {
	// Channel owner, who has created and funded the actor - the sender
	From addr.Address
	// Recipient of payouts from channel
	To addr.Address

	// Amount successfully redeemed through the payment channel, paid out on `Collect()`
	ToSend abi.TokenAmount

	// Height at which the channel can be `Collected`
	SettlingAt abi.ChainEpoch
	// Height before which the channel `ToSend` cannot be collected
	MinSettleHeight abi.ChainEpoch

	// Collections of lane states for the channel, maintained in ID order.
	LaneStates []*LaneState
}
```

Before continuing with the details of the Payment Channel and its components and features, it is worth defining a few terms.

- Voucher: a signed message from the creator of the channel (sender) to the recipient that updates the channel balance.

> - Redeeming a voucher: the voucher is submitted on-chain by the channel's recipient, but does not trigger movement of funds from the channel to the recipient's account. This action does not incur transaction/gas costs.

Reviewer 2: Sorry you're right I meant to say we have lane merging (create a new voucher whose value is the sum of the best vouchers in several lanes).
In any case I believe that anything that gets submitted on chain incurs a gas cost, so the sentence "This action does not incur transaction/gas costs." would need to be changed.

Redeeming a voucher is not transferring funds from the payment channel to the recipient's account. Instead, redeeming a voucher denotes the fact that some service worth of `voucher_val` has been provided by the recipient of the payment channel to the sender. It is not until the voucher is _collected_ that the funds are dispatched to the provider's account.

This is the structure of the voucher:

```go
// A voucher is sent by `From` to `To` off-chain in order to enable
// `To` to redeem payments on-chain in the future
type SignedVoucher struct {
	// ChannelAddr is the address of the payment channel this signed voucher is valid for
	ChannelAddr addr.Address
	// TimeLockMin sets a min epoch before which the voucher cannot be redeemed
	TimeLockMin abi.ChainEpoch
	// TimeLockMax sets a max epoch beyond which the voucher cannot be redeemed
	// TimeLockMax set to 0 means no timeout
	TimeLockMax abi.ChainEpoch
	// (optional) The SecretPreImage is used by `To` to validate
	SecretPreimage []byte
	// (optional) Extra can be specified by `From` to add a verification method to the voucher
	Extra *ModVerifyParams

> 	// Specifies which lane the Voucher merges into (will be created if does not exist)

Reviewer 2: You're right it's lane merging, not voucher merging.
In any case it may be worth using a different word here as lane merging is a different thing from a voucher being added to a lane. A voucher with a higher nonce supersedes a voucher with a lower nonce in the same lane.

Once their transactions have completed, either party can choose to `Settle` (i.e., close) the channel. There is a 12hr period after `Settle` during which the recipient can submit any outstanding vouchers. Once the vouchers are submitted, the recipient can then `Collect` the `ToPay` amount from the channel. The channel sender (`From` address) will be refunded the remaining balance in the channel (if any).

## Lanes

> In addition, payment channels in Filecoin can be split into `lane`s created as part of updating the channel state with a payment `voucher`. Each lane has an associated `nonce` and amount of tokens it can be `redeemed` for. Lanes can be thought of as transactions for several different services provided by the channel recipient to the channel sender. The `nonce` plays the role of a sequence number of vouchers within a given lane, where a voucher with a higher nonce replaces, but does not invalidate a voucher with a lower nonce - the recipient can still redeem or collect a lower-nonce voucher.

Reviewer 2: I believe that submitting a voucher with a higher nonce does actually supersede a voucher with a lower nonce.
So if the client sends (nonce, value) (1,5) then (2,10), the miner can submit (1,5) then (2,10), but cannot submit (2,10) then (1,5).
Note that the lane value = the voucher in the lane with the highest nonce (not the sum of all voucher values)

Payment channel lanes allow for a lot of accounting between parties to be done off-chain and reconciled via single updates to the payment channel. The multiple lanes enable two parties to use a single payment channel to adjudicate multiple independent sets of payments.

> One example of such accounting is *merging of lanes*. When a pair of channel sender-recipient nodes have a payment channel established between them with many lanes, the channel recipient will have to pay gas cost for each one of the lanes in order to `Collect` funds. Merging of lanes allow the recipient to send a "merge" request to the channel sender to request merging of (some of the) lanes and consolidate the funds. This way, the recipient can reduce the overall gas cost. As an incentive for the channel sender to accept the merge lane request, the channel recipient can ask for lower fees to balance out the gas cost. For instance, if the recipient has collected vouchers worth of 10 FIL from two lanes, say 5 from each, and the gas cost of collecting these funds is 2, then it can ask for 9 from the creator if the latter accepts to merge the two lanes. This way, the channel sender pays less overall for the services it received and the channel recipient pays less gas cost to collect the funds for the services they provided.

Reviewer 2: ```suggestion
One example of such accounting is *merging of lanes*. When a pair of channel sender-recipient nodes have a payment channel established between them with many lanes, the channel recipient will have to pay gas cost for each one of the lanes in order to `Collect` funds. Merging of lanes allow the recipient to send a "merge" request to the channel sender to request merging of (some of the) lanes and consolidate the funds. This way, the recipient can reduce the overall gas cost. As an incentive for the channel sender to accept the merge lane request, the channel recipient can ask for a lower total value to balance out the gas cost. For instance, if the recipient has collected vouchers worth of 10 FIL from two lanes, say 5 from each, and the gas cost of submitting the vouchers for these funds is 2, then it can ask for 9 from the creator if the latter accepts to merge the two lanes. This way, the channel sender pays less overall for the services it received and the channel recipient pays less gas cost to submit the voucher for the services they provided.
```

title: Payment Channels
weight: 2
bookCollapseSection: true
dashboardAudit: 1
dashboardState: Stable
dashboardInterface: wip
---

# Payment Channels
---

Payment Channels are used in the Filecoin [Retrieval Market](retrieval_market) to enable efficient off-chain payments and accounting between parties for what is expected to be a series of microtransactions, as these occur during data retrieval.

> In particular, given that there is no proving method provided for the act of sending data from a provider (miner) to a client, there is no trust anchor between the two. Therefore, in order to avoid mis-behaviour, Filecoin is making use of payment channels in order to realise a step-wise data transfer<->payment relationship between the provider and the recipient of data. Clients issue requests for data that the miner is responding to. The miner is entitled to ask for interim payments, the data-oriented interval for which is agreed in the Deal phase. In order to facilitate this process, the Filecoin client is creating a payment channel once the provider has agreed on the proposed deal. The client should also lock in the payment channel monetary value equal to the one needed for retrieval of the entire block of data requested. Every time a provider is completing transfer of the pre-specified amount of data, they can request a payment. The client is responding to this payment with a voucher  which the provider can redeem (immediately or later). Until the voucher is redeemed no monetary value is withdrawn from the payment channel.

Reviewer 2: ```suggestion
In particular, given that there is no proving method provided for the act of sending data from a provider (miner) to a client, there is no trust anchor between the two. Therefore, in order to avoid mis-behaviour, Filecoin makes use of payment channels in order to realise a step-wise data transfer<->payment relationship between the provider (miner) and the recipient of data (client). Clients issue requests for data that the miner responds to. The miner is entitled to ask for interim payments. The data-oriented interval between payments is agreed to in the Deal phase. Once the provider has agreed on the proposed deal, the Filecoin client creates a payment channel. The client locks monetary value in the payment channel equal to the amount needed to retrieve the entire block of data requested. Each time a provider completes transfer of the pre-specified amount of data, they can request a payment. The client responds to this request with a voucher which the provider can redeem (immediately or later). Until the voucher is redeemed no monetary value is withdrawn from the payment channel.
```

Note that the following provides a high-level overview of payment channels as these are implemented in Filecoin and an accompanying interface. The lotus implementation of [vouchers](https://github.com/filecoin-project/lotus/blob/master/chain/types/voucher.go) and [payment channels](https://github.com/filecoin-project/lotus/blob/master/paychmgr/paych.go) are also good references.

You can also read more about the [Filecoin payment channel actor interface](payment_channel_actor).

In short, the payment channel actor can be used to open long-lived, flexible payment channels between users. Each channel can be funded by adding to their balance. 

> The goal of the payment channel actor is to enable a series of off-chain microtransactions to be reconciled on-chain at a later time with fewer messages. Accordingly, the expectation is `From` will send to `To` vouchers of successively greater `Value` and increasing `Nonce`s. When they choose to, `To` can `Update` the channel to update the balance available `ToSend` to them in the channel, and can choose to `Collect` this balance at any time (incurring a gas cost).

Reviewer 2: It may be helpful to include the voucher struct here, so it's clear what `From`, `ToSend` etc are referring to

## Vouchers

Traditionally, in order to transact through a Payment Channel, the payment channel parties send to each other signed messages that update the balance of the channel. In Filecoin, these signed messages are called _vouchers_.

Throughout the interaction between the two parties, the channel sender (`From` address) is sending vouchers to the recipient (`To` address). The `Value` included in the voucher indicates the value available for the receiving party to _redeem_. The `Value` is based on the service that the _payment channel recipient_ has provided to the _payment channel sender_. Either the _payment channel recipient_ or the _payment channel sender_ can `Update` the balance of the channel and the balance `ToSend` to the _payment channel recipient_ (using a voucher), but the `Update` (i.e., the voucher) has to be accepted by the other party before funds can be collected. Furthermore, the voucher has to be redeemed by the opposite party from the one that issued the voucher. The _payment channel recipient_ can choose to `Collect` this balance at any time incurring the corresponding gas cost.

> Redeeming a voucher is not transferring funds from the payment channel to the recipient's account. Instead, redeeming a voucher denotes the fact that some service worth of `Value` has been provided by the payment channel recipient to the payment channel sender. It is not until the voucher is _collected_ that the funds are dispatched to the provider's account.

Reviewer: ```suggestion
Redeeming a voucher is not transferring funds from the payment channel to the recipient's account. Instead, redeeming a voucher denotes the fact that some service worth of `Value` has been provided by the payment channel recipient to the payment channel sender. It is not until the whole payment channel is _collected_ that the funds are dispatched to the provider's account.
```
